cpp/src/merge/merge.cu

```diff
@@ -353,76 +354,154 @@ std::unique_ptr<column> column_merger::operator()<cudf::dictionary32>(column_vie
-                                                          right_col);
-        v_merged_cols.emplace_back(std::move(merged));
-      }
+struct merge_queue_item {
+  table_view view_;
+  table_ptr_type table_;
+  // Priority is a separate member to ensure that moving from an object 
+  // does not change its priority (which would ruin the queue invariant)
+  cudf::size_type priority_ = 0;
+ 
+  void set_priority() {
+    // Smallest tables have the highest priority
+    priority_ = -view_.num_rows();
+  }
+ 
+  merge_queue_item(table_view const& view, table_ptr_type&& table):
+  view_(view), 
+  table_(std::move(table)) {
+    set_priority();
+  }
+ 
+  bool operator<(merge_queue_item const& other) const {
+      return priority_ < other.priority_;
+  }
+};
 
-    return std::make_unique<cudf::experimental::table>(std::move(v_merged_cols));
+// Helper function to ensure that moving out of the priority_queue is "atomic"
+template <typename T>
+T top_and_pop(std::priority_queue<T>& q){
+  auto moved = std::move(const_cast<T&>(q.top()));
+  q.pop();
+  return moved;
 }
+ 
+table_ptr_type merge(std::vector<table_view> const& tables_to_merge,
+                     std::vector<cudf::size_type> const& key_cols,
+                     std::vector<cudf::order> const& column_order,
+                     std::vector<cudf::null_order> const& null_precedence,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream = 0) {
+    if (tables_to_merge.empty()) {
+      return std::make_unique<cudf::experimental::table>();
+    }
+ 
+    auto const& first_table = tables_to_merge.front();
+    auto const n_cols = first_table.num_columns();
+ 
+    for (auto const& table: tables_to_merge) {
+      CUDF_EXPECTS(n_cols == table.num_columns(),
+                   "Mismatched number of columns");
+      CUDF_EXPECTS(cudf::have_same_types(first_table, table),
+                   "Mismatched column types");
+    }
+    
+    CUDF_EXPECTS(!key_cols.empty(),
+                 "Empty key_cols");
+    CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(n_cols), 
+                 "Too many values in key_cols");
+ 
+    CUDF_EXPECTS(key_cols.size() == column_order.size(),
+                 "Mismatched size between key_cols and column_order");
+ 
+    // A queue of (table view, table) pairs
+    std::priority_queue<merge_queue_item> merge_queue;
+    // The table pointer is null if we do not own the table (input tables)
+    std::for_each(tables_to_merge.begin(), 
+                  tables_to_merge.end(), 
+                  [&](auto const& table) { 
+                    if (table.num_rows() > 0)
+                      merge_queue.emplace(table, table_ptr_type()); 
+                  });
+
+    // If there is only one non-empty table_view, return its copy
+    if (merge_queue.size() == 1) {
+      return std::make_unique<cudf::experimental::table>(merge_queue.top().view_);
```

Comment:
This is a sneaky edge case, since `.table_` would be null. I suppose this could be handled earlier by checking if `tables_to_merge` is size 1.

Reply:
This is not the same as `tables_to_merge` size of 1. There could be one or more empty table that was not placed in the queue.

Reply:
Oh I see, you handle empty input tables by discarding them. Looks good